aptos-move/framework/aptos-framework/sources/stake.move

```diff
@@ -1000,33 +1003,45 @@ module aptos_framework::stake {
         );
     }
 
-    /// Mint rewards corresponding to current epoch's `voting_power` and `num_successful_votes`.
+    /// Calculate the rewards amount.
+    fun calculate_rewards_amount(
+        stake_amount: u64,
+        num_successful_proposals: u64,
+        num_total_proposals: u64,
+        rewards_rate: u64,
+        rewards_rate_denominator: u64,
+    ): u64 {
+        // The rewards amount is equal to (stake amount * rewards rate * performance multiplier).
+        // We do multiplication in u128 before division to avoid the overflow and minimize the rounding error.
+        let rewards_numerator = (stake_amount as u128) * (rewards_rate as u128) * (num_successful_proposals as u128);
+        let rewards_denominator = (rewards_rate_denominator as u128) * (num_total_proposals as u128);
```

Comment:
This leaves a chance for an overflow. (we would need u192)

I am thinking whether it is worth doing it this way, vs:
let partial = (stake_amount as u128)*  (num_successful_proposals as u128) / (num_total_proposals as u128)

and then
partial * (rewards_rate as u128) / (rewards_rate_denominator as u128) 

(or change the order for rewards_rate and proposal)

we have no more overflow issues, and might have some imprecision - but due to min_stake_amount being large - does it matter?

Reply:
> This leaves a chance for an overflow. (we would need u192)
Yes, it ideally needs `u192` for the multiplication of three `u64`s. However, @movekevin mentioned that `(stake_amount as u128) * (rewards_rate as u128) * (num_successful_proposals as u128)` would not easily overflow in practice because each multiplicand has a certain upper bound (e.g., stake amount < the market cap, ...).

Reply:
Here's my mental math: (Sentences are a bit choppy because this is copied from Slack messages)

Our initial aptos coin supply is only 1B. So 10^17 units total including decimals of 8. The total stake of each validator is potentially restricted by max stake. Which we might set to 100M or a multiple of that, which is 10^16. Rewards numerator can be 10^9 at most (see vm-genesis/lib.rs). So total is 10^16 * 10^9 * 10^5 = 10^30 at most. 2^128 unsigned can store up to 10^38. So there's still decent headroom with u128.